src/client.rs

```diff
@@ -556,18 +681,94 @@ impl QdrantClient {
         let collection_name = collection_name.to_string();
         let collection_name_ref = collection_name.as_str();
         let ordering_ref = ordering.as_ref();
-
         Ok(self
             .with_points_client(|mut points_api| async move {
-                let result = points_api
+                Ok(points_api
                     .upsert(UpsertPoints {
                         collection_name: collection_name_ref.to_string(),
                         wait: Some(block),
-                        points: points.to_owned(),
+                        points: points.to_vec(),
                         ordering: ordering_ref.cloned(),
                     })
-                    .await?;
-                Ok(result.into_inner())
+                    .await?
+                    .into_inner())
+            })
+            .await?)
+    }
+
+    /// Update or insert points into the collection, splitting in chunks.
+    /// If points with given ID already exist, they will be overwritten.
+    /// This method does *not* wait for completion of the operation, use
+    /// [`upsert_points_batch_blocking`] for that.
+    pub async fn upsert_points_batch(
+        &self,
+        collection_name: impl ToString,
+        points: Vec<PointStruct>,
+        ordering: Option<WriteOrdering>,
+        chunk_size: usize,
+    ) -> Result<PointsOperationResponse> {
+        self._upsert_points_batch(collection_name, &points, false, ordering, chunk_size)
+            .await
+    }
+
+    /// Update or insert points into the collection, splitting in chunks and
+    /// waiting for completion of each.
+    /// If points with given ID already exist, they will be overwritten.
+    pub async fn upsert_points_batch_blocking(
+        &self,
+        collection_name: impl ToString,
+        points: Vec<PointStruct>,
+        ordering: Option<WriteOrdering>,
+        chunk_size: usize,
+    ) -> Result<PointsOperationResponse> {
+        self._upsert_points_batch(collection_name, &points, false, ordering, chunk_size)
+            .await
+    }
+
+    #[inline]
+    async fn _upsert_points_batch(
+        &self,
+        collection_name: impl ToString,
+        points: &[PointStruct],
+        block: bool,
+        ordering: Option<WriteOrdering>,
+        chunk_size: usize,
+    ) -> Result<PointsOperationResponse> {
```

Comment:
As a simplification, you could start right away by testing `if points.len() < chunk_size`.
If it is the case, simply forward the call to `_upsert_points` otherwise the batching can take place.